k2/csrc/ragged_ops.h

```diff
@@ -804,18 +822,99 @@ RaggedShape RemoveEmptyListsAxis0(RaggedShape &src_shape,
-                   src.values[renumbering.New2Old()]);
+Ragged<T> SubsampleRagged(Ragged<T> &src, Renumbering &renumbering,
+                          int32_t axis = -1,
+                          Array1<int32_t> *elems_new2old = nullptr) {
+  Array1<int32_t> tmp;
+  if (elems_new2old == nullptr)
+    elems_new2old = &tmp;
+  RaggedShape shape = SubsampleRaggedShape(src.shape, renumbering,
+                                           axis, elems_new2old);
+  return Ragged<T>(shape, src.values[*elems_new2old]);
 }
 
+/*
+  This function creates a Renumbering object that can be used to obtain subsets
+  of ragged arrays via SubsampleRaggedShape().  It implements beam pruning as
+  used in pruned Viterbi search and similar algorithms, where there is both a
+  beam and a max-active (`max_elems`) constraint.  T will probably be float or
+  double, interpreted as a "positive-is-better" sense, i.e. as scores.
+
+   @param [in] src  The ragged object to be subsampled.
+   @param [in] axis  The axis to be subsampled, must satisfy
+              0 <= axis < src.NumAxes().  The axis before `axis`, if axis < 0,
+              will be interpreted as a "batch" axis.
+   @param [in] beam  The main pruning beam.  The sub-lists of elements on axis
+              `axis` will be removed if their maximum element (or the element
+              itself, if axis + 1 == src.NumAxes()) is less than
+              this_best_elem - beam, where this_best_elem
+              is the maximum element taken over axis `axis-1` (or over the
+              entire array, if axis == 0).   Think of axis `axis-1`, if
+              present, as the "batch" axis, and axis `axis` as the axis that we
+              actually remove elements or sub-lists on.  Empty sub-lists on axis
+              `axis` will always be pruned, as their score would be treated
+              as -infinity.
+   @param [in] max_elems  If max_elems > 0, it is the maximum number of sub-lists
+              or elements that are allowed within any sub-list on axis `axis-1`
+              (or the maximum number of top-level sub-lists after subsampling,
+              if axis == 0).  We keep the best ones, but behavior in case of ties is
+              undefined (TODO: check whether SortSublists() is a stable sort, and
+              change this doc if it is).  If max_elems <= 0, there is no such
+              constraint.
+    @return  Returns the renumbering object to be used to actually
+             prune/subsample the specified axis.
+
+   Example:
+      PruneRagged([ [0 -1 -2 -3], [ -10, -20 ], [ ] ], 1, 5.0, 3)
+    would create a Renumbering object that would prune the
+    ragged tensor to [ [0 -1 -2], [ -10 ], [ ] ]
+
+      PruneRagged([ [0 -1 -2 -3], [ -10, -20 ], [ ] ], 0, 5.0, 0)
+    would create a Renumbering object that would prune the
+    ragged tensor to [ [0 -1 -2 -3] ]
+
+
+  TODO: don't forget to change subsample->subset when we rename
```

Comment:
@danpovey  Do you mean we change the name of original subsample functions to subset (i.e. SubsetRaggedShape, SubsetRagged)?

Reply:
Yes.  Subsample sounds too much like it's regular.